File: CPP/ByRound/1693/1693_B_Fake_Plastic_Trees.cpp
// Problem: CF 1693 B - Fake Plastic Trees
// https://codeforces.com/contest/1693/problem/B

/*
Purpose: 
This code solves the problem of finding the minimum number of operations to set values on vertices of a tree such that each vertex v has a value between l[v] and r[v]. Each operation allows selecting a vertex v and adding a non-decreasing array of non-negative integers to all vertices on the path from root to v.

Algorithm:
- We process nodes in reverse topological order (from leaves to root)
- For each node, we compute how much we need to add to satisfy the constraints
- The key idea is to use a difference array approach on the tree where we propagate the required increments up the tree
- We count how many operations are needed by checking if the current node's excess (d[i]) is less than l[i]

Time Complexity: O(n) per test case
Space Complexity: O(n) for storing parent info, ranges, and difference array

*/
#include <stdio.h>
#include <algorithm>

using namespace std;
#define ll long long
int n, p[200100], ans;
ll l[200100], r[201000], d[200100];

void sol() {
  ans = 0;
  scanf("%d", &n);
  for (int i = 2; i <= n; i++)
    scanf("%d", &p[i]);  // Read parent of each node
  for (int i = 1; i <= n; i++)
    scanf("%lld%lld", &l[i], &r[i]), d[i] = 0;  // Read ranges and initialize diff array
  
  for (int i = n; i; i--) {  // Process nodes in reverse order (leaf to root)
    if (d[i] < l[i]) {  // If current increment is not enough to satisfy lower bound
      d[i] = r[i];  // Set to upper bound (as it's the maximum we allow)
      ans++;  // Increment operation count
    }
    d[i] = min(d[i], r[i]);  // Ensure we don't exceed the upper bound
    
    if (i > 1)  // If not the root
      d[p[i]] += d[i];  // Propagate the increment to parent
  }
  
  printf("%d\n", ans);  // Output result
}

int main() {
  int T;
  scanf("%d", &T);
  while (T--)
    sol();
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByTier/B/1693_B_Fake_Plastic_Trees.py
# Problem: CF 1693 B - Fake Plastic Trees
# https://codeforces.com/contest/1693/problem/B

"""
Algorithm: Tree Processing with Greedy Approach
Techniques: 
- Post-order traversal of tree
- Greedy assignment of values to minimize operations
- Difference array concept on tree

Time Complexity: O(n) per test case, where n is the number of nodes
Space Complexity: O(n) for storing tree structure and auxiliary arrays

This solution processes the tree in reverse topological order (post-order),
assigning values greedily to minimize the number of operations needed.
It uses a lazy propagation-like technique where we track the "excess" 
values that need to be pushed up the tree, and count how many times 
we must "activate" a node to satisfy its constraints.
"""

import sys

input = lambda: sys.stdin.readline()
for _ in range(int(input())):
    n = int(input())
    p = [0, 0] + list(map(int, input().split()))  # Parent array (1-indexed)
    l, r, now, res = (
        [0 for i in range(n + 1)],  # Lower bounds
        [0 for i in range(n + 1)],  # Upper bounds
        [0 for i in range(n + 1)],  # Current excess values
        0,  # Result counter (number of operations)
    )
    for i in range(n):
        l[i + 1], r[i + 1] = map(int, input().split())  # Read l[i], r[i]
    
    # Process nodes in reverse order (post-order traversal)
    for i in range(n, 0, -1):
        # If current excess is less than the required lower bound,
        # we must perform an operation to raise it to the upper bound
        if now[i] < l[i]:
            now[i] = r[i]  # Set to upper bound to satisfy constraint
            res += 1  # Increment operation counter
        
        # Propagate excess to parent (lazy propagation)
        now[p[i]] += min(now[i], r[i])  # Add the minimum of current excess and upper bound
    
    print(res)


# https://github.com/VaHiX/CodeForces/